dns.go panic with root query response

I found a panic with a root query response.  Sanitized pcap attached.

[rootqueryresponse_sanitized_payload.pcap.zip](https://github.com/google/gopacket/files/655186/rootqueryresponse_sanitized_payload.pcap.zip)


panic: runtime error: slice bounds out of range

goroutine 12 [running]:
panic(0x7a0660, 0xc420012120)
	/usr/local/go/src/runtime/panic.go:500 +0x1a1
github.com/google/gopacket/layers.decodeName(0xc420073d6a, 0x1fc, 0x1fc, 0x4c, 0xc4201583a8, 0x1, 0x158, 0x2c0, 0x0, 0xc421124840, ...)
	/home/danm/work/src/github.com/google/gopacket/layers/dns.go:587 +0x6c2
github.com/google/gopacket/layers.(*DNSResourceRecord).decode(0xc421124998, 0xc420073d6a, 0x1fc, 0x1fc, 0x4c, 0x7f26c9c5e590, 0xc4210fd680, 0xc4201583a8, 0x4c, 0x0, ...)
	/home/danm/work/src/github.com/google/gopacket/layers/dns.go:665 +0x6f
github.com/google/gopacket/layers.(*DNS).DecodeFromBytes(0xc420158300, 0xc420073d6a, 0x1fc, 0x1fc, 0x7f26c9c5e590, 0xc4210fd680, 0x1000, 0x1000)
	/home/danm/work/src/github.com/google/gopacket/layers/dns.go:355 +0x74f
github.com/google/gopacket/layers.decodeDNS(0xc420073d6a, 0x1fc, 0x1fc, 0xc1cc80, 0xc4210fd680, 0x0, 0xc)
	/home/danm/work/src/github.com/google/gopacket/layers/dns.go:284 +0xb7
github.com/google/gopacket.DecodeFunc.Decode(0x864bc0, 0xc420073d6a, 0x1fc, 0x1fc, 0xc1cc80, 0xc4210fd680, 0x4218fe, 0xc41ff78873)
	/home/danm/work/src/github.com/google/gopacket/decode.go:85 +0x58
github.com/google/gopacket.LayerType.Decode(0x6b, 0xc420073d6a, 0x1fc, 0x1fc, 0xc1cc80, 0xc4210fd680, 0xc421199400, 0x422bc3)
	/home/danm/work/src/github.com/google/gopacket/layertype.go:95 +0xb1
github.com/google/gopacket.(*LayerType).Decode(0xc4211da060, 0xc420073d6a, 0x1fc, 0x1fc, 0xc1cc80, 0xc4210fd680, 0xc4211da060, 0xc4200358f8)
	<autogenerated>:14 +0x84

--- a/layers/dns.go
+++ b/layers/dns.go
@@ -609,7 +609,7 @@ loop:
 		}
 	}
 	if len(*buffer) <= start {
-		return nil, 0, errDNSNameHasNoData
+		return (*buffer)[start:], index + 1, nil
 	}
 	return (*buffer)[start+1:], index + 1, nil
 }

--- a/layers/dns_test.go
+++ b/layers/dns_test.go
@@ -499,7 +499,7 @@ func TestDNSMalformedPacket2(t *testing.T) {
 	}
 }
 
-// testMalformedRootQuery is the packet:
+// testBlankNameRootQuery is the packet:
 //   08:31:18.143065 IP 10.77.0.26.53 > 10.1.0.233.65071: 59508- 0/13/3 (508)
 //   	0x0000:  0055 22af c637 0022 55ac deac 0800 4500  .U"..7."U.....E.
 //   	0x0010:  0218 76b2 4000 7211 7ad2 0a4d 001a 0a01  ..v.@.r.z..M....
@@ -536,7 +536,7 @@ func TestDNSMalformedPacket2(t *testing.T) {
 //   	0x0200:  0004 ca0c 1b21 c058 0001 0001 0000 0e10  .....!.X........
 //   	0x0210:  0004 c629 0004 c078 0001 0001 0000 0e10  ...)...x........
 //   	0x0220:  0004 c024 9411                           ...$..
-var testMalformedRootQuery = []byte{
+var testBlankNameRootQuery = []byte{
 	0x00, 0x55, 0x22, 0xaf, 0xc6, 0x37, 0x00, 0x22, 0x55, 0xac, 0xde, 0xac, 0x08, 0x00, 0x45, 0x00,
 	0x02, 0x18, 0x76, 0xb2, 0x40, 0x00, 0x72, 0x11, 0x7a, 0xd2, 0x0a, 0x4d, 0x00, 0x1a, 0x0a, 0x01,
 	0x00, 0xe9, 0x00, 0x35, 0xfe, 0x2f, 0x02, 0x04, 0xb8, 0xf5, 0xe8, 0x74, 0x81, 0x00, 0x00, 0x01,
@@ -574,10 +574,10 @@ var testMalformedRootQuery = []byte{
 	0x00, 0x04, 0xc0, 0x24, 0x94, 0x11,
 }
 
-func TestMalformedRootQuery(t *testing.T) {
-	p := gopacket.NewPacket(testMalformedRootQuery, LinkTypeEthernet, testDecodeOptions)
+func TestBlankNameRootQuery(t *testing.T) {
+	p := gopacket.NewPacket(testBlankNameRootQuery, LinkTypeEthernet, testDecodeOptions)
 	if err := p.ErrorLayer(); err != nil {
-		t.Error("Error layer on invalid DNS name:", err)
+		t.Error("Error layer on blank DNS name field:", err)
 	}
 }
 

--- a/layers/dns_test.go
+++ b/layers/dns_test.go
@@ -576,10 +576,8 @@ var testMalformedRootQuery = []byte{
 
 func TestMalformedRootQuery(t *testing.T) {
 	p := gopacket.NewPacket(testMalformedRootQuery, LinkTypeEthernet, testDecodeOptions)
-	if errLayer := p.ErrorLayer(); errLayer == nil {
-		t.Error("No error layer on invalid DNS name")
-	} else if err := errLayer.Error(); !strings.Contains(err.Error(), "no dns data found") {
-		t.Errorf("unexpected error message: %v", err)
+	if err := p.ErrorLayer(); err != nil {
+		t.Error("Error layer on invalid DNS name:", err)
 	}
 }
 

